logger/factory.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package logger
+
+import (
+	"fmt"
+	"github.com/uber-common/bark"
+)
+
+// LoggerFactory wraps a bark.Logger and provides a way to create named loggers
+// restricted to a specific Level.
+type LoggerFactory struct {
+	logger bark.Logger
+	cache  map[string]*namedLogger
+}
+
+func NewLoggerFactory(l bark.Logger) *LoggerFactory {
+	return &LoggerFactory{
+		logger: l,
+		cache:  make(map[string]*namedLogger),
+	}
+}
+
+// SetLevel sets the minimum level for a named logger. A named logger emits
+// messages with a level equal to or greater than this level.
+func (lf *LoggerFactory) SetLevel(name string, level Level) error {
```

Comment:
Not really, we want a centralized facility to configure all named loggers. Consider this:
```go
// in some function
xLogger := factory.Logger("x")
// ... use xLogger

// in another function
xLoggerAgain := factory.Logger("y")
// ... use xLoggerAgain
```
With the `SetLevel` method on the factory we can silence both `xLogger` and `xLoggerAgain` from an admin handler:
```go
factory.SetLevel("x", Off)
```

I'd say it would make sense have the `SetLevel` on the logger if we could somehow have a single logger for each module (something like `Ringpop.Gossip`, `Ringpop.Membership`, etc.). Unfortunately this won't work because of how extra metadata is built using `WithField` and `WithFields`.